fabric-enchantment-api-v1/src/main/java/net/fabricmc/fabric/mixin/enchantment/EnchantmentHelperMixin.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.enchantment;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Unique;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+
+import net.minecraft.enchantment.Enchantment;
+import net.minecraft.enchantment.EnchantmentLevelEntry;
+import net.minecraft.enchantment.EnchantmentTarget;
+import net.minecraft.enchantment.EnchantmentHelper;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+
+import net.fabricmc.fabric.api.enchantment.v1.EnchantmentEvents;
+import net.fabricmc.fabric.api.enchantment.v1.FabricEnchantment;
+import net.fabricmc.fabric.api.util.TriState;
+
+/**
+ * A mixin to handle the verification of FabricEnchantments with custom targets.
+ *
+ * @author Vaerian (vaeriann@gmail.com or @Vaerian on GitHub).
+ *
+ * <p>Please contact the author, Vaerian, at the email or GitHub profile listed above
+ * with any questions surrounding implementation choices, functionality, or updating
+ * to newer versions of the game.</p>
+ */
+@Mixin(EnchantmentHelper.class)
+public class EnchantmentHelperMixin {
+	@Unique
+	private static Enchantment currentEnchantment;
+
+	// A simple read/write lock to prevent a data race by ensure that the
+	// value captured from the inject is not modified by the time it gets to the redirect mixin.
+	@Unique
+	private static Lock lock = new ReentrantLock();
```

Comment:
Not sure if it would just be better to use a `ThreadLocal` here and avoid the weird while loops below

Reply:
The trade-off would be probably minimally more memory overhead in exchange for some minor performance gains in the case of two threads trying to access the enchantments (which should also typically never occur), but I did change it and we can see if we like that more.